src/optlang/coinor_cbc_interface.py

```diff
@@ -0,0 +1,816 @@
+
+    def _var_dual(self, name):
+        dual = self.problem.var_by_name('v_' + name).rc
+        return 0. if dual is None else dual
+
+    def _get_primal_values(self):
+        if getattr(self, '_status', '') == 'optimal':
+            return super(Model, self)._get_primal_values()
+        return [0. for _ in self.variables]
+
+    def _get_reduced_costs(self):
+        if getattr(self, '_status', '') == 'optimal':
+            return super(Model, self)._get_reduced_costs()
+        return [0. for _ in self.variables]
+
+    def _update_var_lb(self, name, lb):
+        self.problem.var_by_name('v_' + name).lb = to_float(lb, True)
+
+    def _update_var_ub(self, name, ub):
+        self.problem.var_by_name('v_' + name).ub = to_float(ub, False)
+
+    def _update_var_type(self, name, var_type):
+        self.problem.var_by_name('v_' + name).var_type = var_type
+
+    def _add_variables(self, variables):
+        super(Model, self)._add_variables(variables)
+        for var in variables:
+            # TODO: may need to handle obj, column options
+            self.problem.add_var(name='v_' + var.name,
+                                 var_type=_VTYPE_TO_MIP_VTYPE[var.type],
+                                 lb=to_float(var.lb, True),
+                                 ub=to_float(var.ub, False))
+
+    def _remove_variables(self, variables):
+        # TODO: optimization for removing all variables?
+        if self.objective is not None:
+            self.objective._changed_expression.update({var: 0 for var in variables})
+        mip_vars = []
+        for var in variables:
+            name = var.name
+            del self._variables_to_constraints_mapping[name]
+            var.problem = None
+            del self._variables[name]
+            mip_vars.append(self.problem.var_by_name('v_' + name))
+        self.problem.remove(mip_vars)
+
+    def _constr_primal(self, con, is_lb):
+        slack = self.problem.constr_by_name(con.constraint_name(is_lb)).slack
+        return con.lb + slack if is_lb else con.ub - slack
+
+    def _constr_dual(self, con, is_lb):
+        return self.problem.constr_by_name(con.constraint_name(is_lb)).pi
+
+    def _get_constraint_values(self):
+        if getattr(self, '_status', '') == 'optimal':
+            return super(Model, self)._get_constraint_values()
+        return [0. for _ in self.constraints]
+
+    def _get_shadow_prices(self):
+        if getattr(self, '_status', '') == 'optimal':
+            return super(Model, self)._get_shadow_prices()
+        return [0. for _ in self.constraints]
+
+    def _update_constraint_bound(self, con, is_lb):
+        name = con.constraint_name(is_lb)
+        constr = self.problem.constr_by_name(name)
+        constr.rhs = to_float(-1*con.lb if is_lb else con.ub)
+
+    def _expr_to_mip_expr(self, expr):
+        """Parses mip linear expression from expression."""
+        offset, coeffs, _ = parse_optimization_expression(expr)
```

Comment:
Change to `parse_optimization_expression(expr.expand())` to make it work in symengine.

Reply:
`expr.expand()` fails for me since `expr` is an `Objective` or a `Constraint`, which do not have that method. Expanding the internal expression resolved the error at `test_imul_objective`:
```python
    def _expr_to_mip_expr(self, expr):
        """Parses mip linear expression from expression."""
        if hasattr(expr, "expression") and symbolics.USE_SYMENGINE:
            expr._expression = expr.expression.expand()
        offset, coeffs, _ = parse_optimization_expression(expr)
        return offset + mip.xsum(to_float(coef) * self.problem.var_by_name('v_' + var.name)
                                 for var, coef in coeffs.items())
```
Nevertherless, it is still failing because of the dummy comparison at `test_change_objective`:
```
AssertionError: 0.0 + 1.0*ACALD + 1.0*ACALD_reverse_fda2b != 1.0*ACALD + 1.0*ACALD_reverse_fda2b
```

Reply:
Just have to call simplify on the expression first.

Reply:
Awesome, calling simplify (on both sides) makes the test suite pass (lines 335 and 340 of `src/optlang/tests/test_coinor_cbc_interface.py`).

Reply:
Yeah exactly. This is what we use in cobrapy as well to cover sympy and symengine at the same time.

```Python
def same_ex(ex1, ex2):
    """Compare to expressions for mathematical equality."""
    return ex1.simplify() == ex2.simplify()
```

Reply:
That is handy, definitely. This PR can be merged after applying these two small changes, any thoughts @braceal ?

Reply:
Hello, Thanks for the activity on this! It would be great to get it merged. 

I remember needing to update the expression parsing logic. Looking back at the code, I believe there was an edge case when a single constant value was used as an expression. Just wanted to make sure this was checked by someone more familiar with the codebase.

https://github.com/opencobra/optlang/pull/213/files#diff-9be3c58a107490c43a136ec465f971e6ef2f76dd5f0080f687cc62e86ae96b05R83

Reply:
@braceal If you enable the `Allow edits and access to secrets by maintainers` checkmark on the upper right we could try to fix it directly in this PR.

Reply:
@cdiener It says it is already enabled. Are you not able to commit changes?
![image](https://user-images.githubusercontent.com/38300604/110556022-b79bda80-810b-11eb-8d25-85564a1ab062.png)

Reply:
Ups, sorry. Didn't even try. It showed nothing for me so I assumed it's not on but it's just bad UI :sweat_smile:

Reply:
@braceal Your fix to `expression_parsing` looks fine. Good catch.

Reply:
Thanks for reviewing @KristianJensen 

What items are remaining before a merge is possible?

Reply:
Let me finish merging #225 and I will make this two couple of changes to pass the CI.

Reply:
Probably need to merge the CI fixes from the master branch first (we should rename that one btw).

Reply:
Already merged CI fixes to devel but now OSQP is failing for Windows 3.9 because of https://github.com/oxfordcontrol/osqp/issues/240

Reply:
See my comments on the otehr PR and the new one I sent. It's more a version thing. I don't know tox very well though so no idea if it will work.